### PR TITLE
ddl: add partition compatibility for temporary table

### PR DESCRIPTION
**Pull request by Howie59**
### What problem does this PR solve?

Issue Number: see https://github.com/pingcap/tidb/pull/24168

### What is changed and how it works?

Proposal: [temporary table](https://github.com/pingcap/tidb/issues/24169) 

What's Changed:  add partition compatibility for temporary table

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

**Comment by morgo** [comment]
The `check_dev` is failing because of the following error:

```
Files errors.toml and /tmp/errors.toml.before differ
```

Please run `make dev` and commit the new `errors.toml` file.

Otherwise LGTM

File: ddl/error.go [hunk 1]
```diff
@@ -274,4 +274,7 @@ var (
 	ErrUnknownEngine = dbterror.ClassDDL.NewStd(mysql.ErrUnknownStorageEngine)
 
 	errExchangePartitionDisabled = dbterror.ClassDDL.NewStdErr(mysql.ErrUnsupportedDDLOperation, parser_mysql.Message("Exchange Partition is disabled, please set 'tidb_enable_exchange_partition' if you need to need to enable it", nil))
+
+	// ErrPartitionNoTemporary returns when partition at temporary mode
+	ErrPartitionNoTemporary = dbterror.ClassAdmin.NewStd(mysql.ErrPartitionNoTemporary)
```
Comment (djshow832) [inline ddl/error.go:279]: Why `ClassAdmin`?

File: ddl/table.go [hunk 2]
```diff
@@ -56,6 +56,11 @@ func onCreateTable(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
 		job.State = model.JobStateCancelled
 		return ver, errors.Trace(err)
 	}
+	if tbInfo.Partition != nil && tbInfo.Partition.Enable && tbInfo.TempTableType == model.TempTableGlobal {
```
Comment (djshow832) [inline ddl/table.go:59]:
```suggestion
	if tbInfo.Partition != nil {
```
No need to check `Enable`. It's a historical problem.
Besides, it also doesn't support local temporary tables in the future.

File: ddl/db_partition_test.go [hunk 3]
```diff
@@ -3385,3 +3386,14 @@ func (s *testIntegrationSuite7) TestPartitionListWithNewCollation(c *C) {
 	str := tk.MustQuery(`desc select * from t11 where a = 'b';`).Rows()[0][3].(string)
 	c.Assert(strings.Contains(str, "partition:p0"), IsTrue)
 }
+
+func (s *testIntegrationSuite7) TestAddTableWithPartition(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("use test;")
+	tk.MustExec("drop table if exists global_partition_table;")
+	tk.MustGetErrCode("create global temporary table global_partition_table (a int, b int) partition by hash(a) partitions 3 ON COMMIT DELETE ROWS;", errno.ErrPartitionNoTemporary)
```
Comment (djshow832) [inline ddl/db_partition_test.go:3394]: Please also test range/list partition.

File: ddl/db_partition_test.go [hunk 4]
```diff
@@ -3385,3 +3386,27 @@ func (s *testIntegrationSuite7) TestPartitionListWithNewCollation(c *C) {
...
+	tk.MustExec("drop table if exists global_partition_table;")
+	_, err := tk.Exec("create table partition_table (a int, b int) partition by hash(a) partitions 3;")
+	c.Assert(err, IsNil)
+	tk.MustExec("drop table if exists partition_table;")
+	_, err = tk.Exec(`CREATE TABLE partition_range_table (c1 smallint(6) NOT NULL, c2 char(5) DEFAULT NULL) PARTITION BY RANGE ( c1 ) (
```
Comment (djshow832) [inline ddl/db_partition_test.go:3399]: I mean checking range/list partition is not compatible with temporary tables.

File: ddl/db_partition_test.go [hunk 5]
```diff
@@ -3385,3 +3386,30 @@ func (s *testIntegrationSuite7) TestPartitionListWithNewCollation(c *C) {
...
 	c.Assert(strings.Contains(str, "partition:p0"), IsTrue)
 }
+
+func (s *testIntegrationSuite7) TestAddTableWithPartition(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("use test;")
+	tk.MustExec("drop table if exists global_partition_table;")
+	tk.MustGetErrCode("create global temporary table global_partition_table (a int, b int) partition by hash(a) partitions 3 ON COMMIT DELETE ROWS;", errno.ErrPartitionNoTemporary)
+	tk.MustExec("drop table if exists global_partition_table;")
+	_, err := tk.Exec("create table partition_table (a int, b int) partition by hash(a) partitions 3;")
```
Comment (djshow832) [inline ddl/db_partition_test.go:3396]: Drop `partition_table` before creating.

**Pull request merged**